### PR TITLE
[FEATURE] Allow concurrent calls to bulk_contains

### PR DESCRIPTION
**Pull request by eseiler**
Resolves https://github.com/seqan/product_backlog/issues/78

### CODE REVIEW

**Comment by marehr** [comment]
@eseiler Just to be sure, should I already review it?

**Comment by eseiler** [comment]
> @eseiler Just to be sure, should I already review it?

IMO yes

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 1]
```diff
@@ -101,6 +101,9 @@ struct bin_index : public detail::strong_type<size_t, bin_index, detail::strong_
  * length `b` starting at the positions indicated by the hash functions. The bitwise AND of these sub-bitvectors yields
  * the binningvector, a bitvector of length `b` where the `i`'th bit indicates set membership in the `i`'th bin.
  *
+ * ### Querying
+ * To query the Interleaved Bloom Filter for a value use the seqan3::bulk_agent.
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:105]: This comment is outdated right?

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 2]
```diff
@@ -402,46 +353,21 @@ class interleaved_bloom_filter
...
+    /*!\brief Returns seqan3::interleaved_bloom_filter::bulk_agent to be used for lookup.
+     * \attention Calling seqan3::interleaved_bloom_filter::increase_bin_number_to invalidates all
+     * seqan3::interleaved_bloom_filter::bulk_agent constructed for this Interleaved Bloom Filter.
      *
      * \details
      *
      * ### Example
      *
-     * \include test/snippet/search/dream_index/interleaved_bloom_filter_bulk_contains.cpp
+     * \include test/snippet/search/dream_index/bulk_agent_constructor.cpp
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:364]: The name of the snippet is a bit misleading, but it is okay for me :)

Maybe `bulk_agent_construction.cpp`

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 3]
```diff
@@ -114,66 +118,18 @@ struct bin_index : public detail::strong_type<size_t, bin_index, detail::strong_
...
-        friend class interleaved_bloom_filter;
-        using sdsl::bit_vector::resize;
-        using sdsl::bit_vector::set_int;
-    };
-
 private:
     //!\cond
     template <data_layout data_layout_mode>
     friend class interleaved_bloom_filter;
+    class bulk_agent;
```
Comment (smehringer) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:132]:
```suggestion
    friend class interleaved_bloom_filter;
    
    class bulk_agent;
```

**Comment by eseiler** [comment]
BTW, since we are not sure about the naming, we can also kinda postpone this. 
I can change the names again in https://github.com/seqan/seqan3/pull/1911 the latest, were we will also have the use case of another agent for the binning_directory that can be used for counting.

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 4]
```diff
@@ -525,11 +451,157 @@ class interleaved_bloom_filter
...
+ */
+template <data_layout data_layout_mode>
+class interleaved_bloom_filter<data_layout_mode>::bulk_agent
+{
+private:
+    //!\brief The type of the augmented seqan3::interleaved_bloom_filter.
+    using ibf_t = interleaved_bloom_filter<data_layout_mode>;
+
+    //!\brief A pointer to the augmented seqan3::interleaved_bloom_filter.
+    ibf_t const * ibf_ptr;
```
Comment (smehringer) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:475]:
```suggestion
    ibf_t const * ibf_ptr{nullptr};
```
or is this equivalent?

File: test/snippet/search/dream_index/bulk_agent_bulk_contains.cpp [hunk 5]
```diff
@@ -11,6 +11,7 @@ int main()
     // Query the Interleaved Bloom Filter. Note that there may be false positive results!
     // A `1` at position `i` indicates the (probable) presence of the query in bin `i`.
     // Capture the result by reference to avoid copies.
-    auto & result = ibf.bulk_contains(712);
+    auto agent = ibf.agent();
+    auto & result = agent.bulk_contains(712);
     seqan3::debug_stream << result << '\n'; // prints [0,0,0,1,0,0,0,0,0,0,0,0]
```
Comment (smehringer) [inline test/snippet/search/dream_index/bulk_agent_bulk_contains.cpp:16]:
```suggestion
    auto & result = agent.bulk_contains(712);
    seqan3::debug_stream << result << '\n'; // prints [0,0,0,1,0,0,0,0,0,0,0,0]
    
    // Calling `increase_bin_number_to` invalidates the agent.
    ibf.increase_bin_number_to(seqan3::bin_count{60u});

    // So make sure to construct a new bulk_agent.
    agent = ibf.agent();
```

File: test/snippet/search/dream_index/interleaved_bloom_filter_increase_bin_number_to.cpp [hunk 6]
```diff
@@ -9,9 +9,10 @@ int main()
     ibf.emplace(237, seqan3::bin_index{9u});
 
     ibf.increase_bin_number_to(seqan3::bin_count{18u});
+    auto agent = ibf.agent();
```
Comment (smehringer) [inline test/snippet/search/dream_index/interleaved_bloom_filter_increase_bin_number_to.cpp:12]:
```suggestion
    auto agent = ibf.agent(); // be sure to get the agent after `increase_bin_number_to` as it invalidates all agents!
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp (hunk 4 above)
Comment (eseiler) [reply include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:475]: It's equivalent but I like the explicit version more

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 7]
```diff
@@ -525,11 +452,157 @@ class interleaved_bloom_filter
...
 };
 
+/*!\brief Manages queries for the seqan3::interleaved_bloom_filter.
+ * \attention Calling seqan3::interleaved_bloom_filter::increase_bin_number_to on `ibf` invalidates the bulk_agent.
+ *
+ * \details
+ *
+ * ### Example
+ *
+ * \include test/snippet/search/dream_index/bulk_agent_constructor.cpp
```
Comment (eseiler) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:466]:
```suggestion
 * \include test/snippet/search/dream_index/bulk_agent_construction.cpp
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 8]
```diff
@@ -525,11 +452,159 @@ class interleaved_bloom_filter
         archive(bin_words);
         archive(hash_funs);
         archive(data);
-        result_buffer.resize(bins);
     }
     //!\endcond
 };
 
+/*!\brief Manages queries for the seqan3::interleaved_bloom_filter.
```
Comment (smehringer) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:459]:
```suggestion
/*!\brief Manages membership queries for the seqan3::interleaved_bloom_filter.
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 9]
```diff
@@ -525,11 +452,159 @@ class interleaved_bloom_filter
...
+     */
+    membership_agent() = default; //!< Defaulted.
+    membership_agent(membership_agent const &) = default; //!< Defaulted.
+    membership_agent & operator=(membership_agent const &) = default; //!< Defaulted.
+    membership_agent(membership_agent &&) = default; //!< Defaulted.
+    membership_agent & operator=(membership_agent &&) = default; //!< Defaulted.
+    ~membership_agent() = default; //!< Defaulted.
+
+    /*!\brief Construct a membership_agent for an existing seqan3::interleaved_bloom_filter.
+     * \private
```
Comment (smehringer) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:493]: ?
Why only make the param private? (if that was the intention at all)

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 10]
```diff
@@ -525,11 +452,159 @@ class interleaved_bloom_filter
...
+     * \{
+     */
+    membership_agent() = default; //!< Defaulted.
+    membership_agent(membership_agent const &) = default; //!< Defaulted.
+    membership_agent & operator=(membership_agent const &) = default; //!< Defaulted.
+    membership_agent(membership_agent &&) = default; //!< Defaulted.
+    membership_agent & operator=(membership_agent &&) = default; //!< Defaulted.
+    ~membership_agent() = default; //!< Defaulted.
+
+    /*!\brief Construct a membership_agent for an existing seqan3::interleaved_bloom_filter.
```
Comment (smehringer) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:492]:
```suggestion
    /*!\brief Construct a membership_agent from an seqan3::interleaved_bloom_filter.
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp (hunk 9 above)
Comment (eseiler) [reply include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:493]: This marks the constructor as private, not just the param.

**Pull request merged**